instrumentation/instasarama/propagation.go

```diff
@@ -40,47 +55,75 @@ type ProducerMessageCarrier struct {
 
 // Set implements opentracing.TextMapWriter for ProducerMessageCarrier
 func (c ProducerMessageCarrier) Set(key, val string) {
+	kafkaHeaderFormat := GetKafkaHeaderFormat()
 	switch strings.ToLower(key) {
 	case instana.FieldT:
-		if len(val) > 32 {
-			return // ignore hex-encoded trace IDs longer than 128 bit
-		}
+		if kafkaHeaderFormat == BOTH || kafkaHeaderFormat == BINARY {
+			if len(val) > 32 {
+				return // ignore hex-encoded trace IDs longer than 128 bit
+			}
 
-		traceContext := PackTraceContextHeader(val, "")
-		if i, ok := c.indexOf(fieldCKey); ok {
-			// preserve the trace ID if the trace context header already present
-			existingC := c.Message.Headers[i].Value
-			if len(existingC) >= 16 {
-				copy(traceContext[16:], existingC[16:])
+			traceContext := PackTraceContextHeader(val, "")
+			if i, ok := c.indexOf(fieldCKey); ok {
+				// preserve the trace ID if the trace context header already present
+				existingC := c.Message.Headers[i].Value
+				if len(existingC) >= 16 {
+					copy(traceContext[16:], existingC[16:])
+				}
 			}
+
+			c.addOrReplaceHeader(fieldCKey, traceContext)
 		}
 
-		c.addOrReplaceHeader(fieldCKey, traceContext)
-	case instana.FieldS:
-		if len(val) > 16 {
-			return // ignore hex-encoded span IDs longer than 64 bit
+		if kafkaHeaderFormat == BOTH || kafkaHeaderFormat == STRING {
+			existingT := val
+			valLen := len(val)
+			if valLen < 32 {
+				existingT = strings.Repeat("0", 32-valLen) + val
+			}
+			c.addOrReplaceHeader(fieldTKey, []byte(existingT))
 		}
+	case instana.FieldS:
+		if kafkaHeaderFormat == BOTH || kafkaHeaderFormat == BINARY {
+			if len(val) > 16 {
+				return // ignore hex-encoded span IDs longer than 64 bit
+			}
 
-		traceContext := PackTraceContextHeader("", val)
-		if i, ok := c.indexOf(fieldCKey); ok {
-			// preserve the span ID if the trace context header already present
-			existingC := c.Message.Headers[i].Value
-			if len(existingC) >= 16 {
-				copy(traceContext[:16], existingC[:16])
+			traceContext := PackTraceContextHeader("", val)
+			if i, ok := c.indexOf(fieldCKey); ok {
+				// preserve the span ID if the trace context header already present
+				existingC := c.Message.Headers[i].Value
+				if len(existingC) >= 16 {
+					copy(traceContext[:16], existingC[:16])
+				}
 			}
+
+			c.addOrReplaceHeader(fieldCKey, traceContext)
 		}
 
-		c.addOrReplaceHeader(fieldCKey, traceContext)
+		if kafkaHeaderFormat == BOTH || kafkaHeaderFormat == STRING {
+			c.addOrReplaceHeader(fieldSKey, []byte(val))
```

Comment:
For the `BINARY` case we check the length of the value, do we want to do the same here?